### ISSUE TITLE
Should we provide our own JSON codecs for the JsonSchemas algebra?

### ISSUE THREAD

**Issue by julienrf** [julienrf/endpoints#399]
Following up the discussion below, I plan to do the following tasks:

- [x] Implement JSON encoders for `JsonSchemas` algebra
- [x] Use these JSON encoders in `openapi.JsonSchemas` interpreter, and leverage the fact that we already have a dependency to a JSON library to implement an `Encoder[OpenApi]` (as opposed to the current situation where we use `OpenApiSchemas` with a `JsonSchemas` codec interpreter)
- [x] Fix #185
- [x] Fix #121
- [x] Implement JSON decoders for `JsonSchemas` algebra
- [x] Remove the need to mix a JSON codec interpreter with `JsonSchemaEntities` since we can use the built-in one

----

It started as an anecdotical question and eventually pulled significant design reconsiderations. I’d like to request your feedback.

I was looking at #185 (generalizing enumerations to types other than `String`) and realized that solving this issue would require a significant change. Eventually, we’d need the `openapi` interpreter to not only be able to produce documentation for `JsonSchema[A]` values, but also to be able to **encode `A` values in JSON**. So, it would have to provide an `Encoder[A, Json]`. So far, such encoders were provided by the `playjson` and `circe` interpreters, which delegate to Play-json and Circe the actual implementation of the codecs. The idea was that people could pick their preferred JSON backend between those two.

So, when I realized that solving #185 would require including a JSON encoder, I was tempted to introduce a way for users to provide a `circe` or `playjson` interpreter to the `openapi` interpreter, so that the `openapi` could benefit from their JSON encoders. This would require little work in _endpoints_ but would be a bit complicated/confusing for end-users, I’m afraid.

To keep things simple, I’d rather **implement a JSON encoder in the `openapi` interpreter**. End-users wouldn’t have to do anything, and we could solve #185, as well as #121. And we could also simplify the way we generate the OpenAPI document itself (which is currently defined as a JSON schema, in the `OpenApiSchemas` trait, in which the user has to mix a codec interpreter to produce a concrete `OpenApiEncoder`, as explained [here](http://julienrf.github.io/endpoints/interpreters/openapi.html#endpoints)). So, this would be super cool.

One concern I have with implementing a JSON encoder in the `openapi` interpreter is that it could lead to a “transitive dependencies hell” for end-users. For instance, if I implement the JSON encoder with circe, then people using the `openapi` interpreter would transitively depend on circe, which itself depends on cats. The `openapi` interpreter is typically used in backends/servers, so its means that end-users would have to align their cats version to the one used by the `openapi` interpreter. It also means that end-users might pull both Circe and Play-json in their final application, in case they were using `play-json` in the backend. Therefore, I was thinking of using a JSON library with little dependencies. Maybe mdedetrich/scalajson?

But the story doesn’t stop here.

From providing a JSON encoder to **providing both an encoder and a decoder**, there is just one step. So, should we provide an interpreter of JsonSchemas that produces codecs? I mean, an interpreter that does not introduces a Circe dependency or a Play-json dependency. The `openapi` would just use it to get its JSON encoders. But this would also bring the following additional benefits:

- we would control the format of validation error messages (currently, we have to re-process the errors produced by Play-json or Circe to fit them into our `Invalid` data type, but the final representation [is not great](https://github.com/julienrf/endpoints/blob/4fb381683155654bfa545f027a5f12b23731214b/akka-http/server-playjson/src/test/scala/endpoints/akkahttp/server/playjson/EndpointsJsonSchemaTest.scala#L73)),
- end-users wouldn’t have to deal with the “modularity hell” of _endpoints_, the `akka-http-server-circe`, `akka-http-client-circe`, `akka-http-server-playjson`, etc. these modules wouldn’t anymore be necessary to use `JsonSchemaEntities`.

However, compared to the status quo, we would lose the following capability:

- `JsonSchema[?]` values could not anymore be interoperable with Play-json or Circe (currently, the `circe` interpreter maps `JsonSchema[A]` to something like `io.circe.Codec[A]`, which means that your backend can reuse these codecs at other places (not just for serializing/deserializing HTTP entities), if needed). That being said, we could keep the existing `circe.JsonSchemas` and `playjson.JsonSchemas` interpreters for this (even though we wouldn’t use it to implement `JsonSchemaEntities`).

If we go this way, I would use typelevel/jawn for JSON parsing, AST, and rendering.

To summarize, I see two directions into which we could go:

1. Implement a JSON serializer in the `openapi` interpreter. This would solve #185, #121, and simplify the way we encode the OpenAPI document itself into JSON. I would use mdedetrich/scalajson for this.
2. Or, even implement an interpreter producing JSON codecs for the `JsonSchemas` algebra, and use its encoders in the `openapi` interpreter. This would solve the same issues as 1., but, would simplify even further the usage of `JsonSchemaEntities` and would improve the quality of JSON validation error messages. I would use typelevel/jawn for this.

I have a preference for the 2nd solution.

Any thoughts?

**Comment by Krever** [comment]
I like the 2 approach as well. One thing I don't understand is non-interoperability of the codecs. If I follow correctly, then we will have some `endpoints.Codec[T]` that could be translated (by interop layer) to play `Reads`/`Writes`/`Format` or circe `Encoder`/`Decoder`, right?

**Comment by julienrf** [comment]
> One thing I don't understand is non-interoperability of the codecs

It’s just that currently, the `circe.JsonSchemas` interpreter produces a Circe `Encoder` and `Decoder` that can directly be used by your application to serialize and deserialize values at other places than for HTTP requests and responses. (Same with Play-json)

With a `jawn.JsonSchemas` interpreter used by `JsonSchemaEntities` (for serializing and deserializing HTTP requests and responses), then users could not directly use the produced codecs with Circe or Play-json.

But this interoperability would be recovered by creating an object applying the `circe.JsonSchemas` interpreter to the application’s JSON schema definitions. Or, as you said, by writing a simple interop layer that would translate the jawn codecs to Circe or Play-json codecs.

**Comment by Krever** [comment]
I see. I think that currently "modularity-hell" is the biggest source of complexity in endpoints, so any tries to fix it sounds nice.

**Comment by gmarti** [comment]
I like the 2nd approach too, modularity provides complexity in this case and removing it will help understandability.
If we can keep circe.JsonSchemas :+1: its cool
But it will be more like a "plugin"

**Comment by jchapuis** [comment]
Nice write-up, indeed the second alternative looks promising: probably that native JSON capabilities at the interpretation layer will indeed lead to gains in flexibility and make it more plug-and-play for users, as you explained. Native support for common types e.g. the ones from java.time would also be super useful IMO.
I remember using generated codecs in tests in some instance (can't recall what for, I think for manipulating test data).

**Comment by julienrf** [comment]
I’ve been working on this milestone lately and have split the changes in multiple PR for easier review.

The most important deviation from the prior design decisions is that _endpoints_ now embeds a JSON library (ujson), whereas it used to provide an empty API (so-called algebra interfaces) implemented by forwarding to the specific libraries the end-user was already using (Circe, Akka, etc.).

I think the benefits are nice (not crazy, but still nice!). Beyond the fact that it makes it possible to generalize enumerations (#412) and to embed examples of entities in the documentation (#416), it also reduces the number of modules (#420) and hopefully simplifies the developer experience.

I think this is the way to go and, in the coming days, I will polish the PRs I’ve opened to get them merged. Feel free to comment on them if you have any concerns!
